### ISSUE TITLE
S3 client could not make PUT request with http protocol 

### ISSUE THREAD

**Issue by cyflhn** [aws/aws-sdk-java-v2#1297]
S3 client could not make PUT request if http protocol is used.
## Expected Behavior

The PUT api of S3 client should work well for both http and https protocol. But for http protocol, there is a problem in it. 
## Current Behavior

I use s3 client to upload files to s3. I set a value which is an http url for endpoint. Then when I invoke the API of `s3Client.putObject`, client side is keeping waiting response from server. 
I have checked the source code of s3 API. If http protocol is used, then the whole payload will be signed,including header and body. So during the process of body signing, the whole stream is read. When the client side lauch a PUT request, at first step, the server respond with 100 continue, then client side should send body to the sever, but the stream has been adready consumed. the client is haning until timeout occur.
## Possible Solution

The code of siging the http request body should be modified.
## Your Environment

* AWS Java SDK version used: 
* JDK version used:1.8.0_191
* Operating System and version: centos 7

**Comment by zoewangg** [comment]
What Java SDK version are you using?  
What type of the `InputStream` are you using? Does it support resetting?
Can you provide sample code to reproduce the issue?

**Comment by cyflhn** [comment]
@zoewangg 
The jdk version is 1.8.0_191. The aws s3 sdk version  is 2.5.31.
The code is very simple:
```
 S3ClientBuilder builder = S3Client.builder().region(Region.of(config.getRegion()))
                .credentialsProvider(StaticCredentialsProvider.create(awsBasicCredentials))
                .httpClientBuilder(ApacheHttpClient.builder().connectionTimeout(Duration.ofSeconds(30))
                       .connectionMaxIdleTime(Duration.ofSeconds(30)).socketTimeout(Duration.ofSeconds(60)));
 builder = builder.endpointOverride(new URI("http://xxxxxxx"));
putObjectResponse =
                        s3Client.putObject(putObjectRequest, RequestBody.fromBytes(transferContent.getData()));
```
Maybe there are some missunderstandings here.  I have analyzed the source of aws sdk: the class of `software.amazon.awssdk.auth.signer.AwsS3V4Signer` and `software.amazon.awssdk.auth.signer.internal.AbstractAws4Signer`.
I did not use any `Inputstream`. The inutputstream I mentioned is used in aws s3 sdk, the `Inputstream` represents the stream data of http request.
It's quite easy to reproduce.

**Comment by zoewangg** [comment]
I see. I thought you were using a customized `RequestBody` with an `InputStream`. Anyway, I was not able to reproduce the issue using similar code.

The SDK does sign payload body by default if TLS is not being used, but it would use chunked encoding, which essentially wraps the underlying inputstream (`ByteArrayInputStream` in this case) with `AwsChunkedEncodingInputStream`. If chunked encoding is disabled (enabled by default), the SDK would invoke a new stream by `streamProvider.newStream()` and then buffers the content to calculate the hash of the content. Nevertheless, when the SDK sends out a request, a "new" stream is being used. 

https://github.com/aws/aws-sdk-java-v2/blob/b10264977af301f7a0c97af69e22146b8654d88c/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntity.java#L114

So I think the timeout error you were seeing was caused by something else. If you can enable the [wire logs](https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/java-dg-logging.html) and post the logs (please redact sensitive information), we can take a deeper look.

**Comment by cyflhn** [comment]
@zoewangg 
I did not use a customized `RequestBody`. The `RequestBody` is also from aws sdk. It's `software.amazon.awssdk.core.sync.RequestBody`. 
I have disabled chunked encoding as s3 adapter does not support chunk encoding.
But it is really easy to reproduce. I have modified the code of the method  `isPayloadSigningEnabled 
 software.amazon.awssdk.auth.signer.AwsS3V4Signer`. I totally disabled the payload body signing. Then the problem disappeared.

**Comment by cyflhn** [comment]
@zoewangg 
I have debugged the code again. You said "the SDK would invoke a new stream by streamProvider.newStream() ". It's true. But the value returned by `streamProvider.newStream() ` is an object of `ChecksumCalculatingInputStream`. The internal inputstream in `ChecksumCalculatingInputStream` which is a `ByteArrayInputStream` is not a replicated one.

**Comment by zoewangg** [comment]
Okay,  disabling `chunkedEncodingEnabled`  is the missing part and I'm now able to reproduce the issue.
```
        return S3Client.builder()
                       .region(DEFAULT_REGION)
                       .endpointOverride(URI.create("http://xxx"))
                       .serviceConfiguration(S3Configuration.builder().chunkedEncodingEnabled(false).build())
                       .credentialsProvider(CREDENTIALS_PROVIDER_CHAIN)
```
We will look into it.

**Comment by zoewangg** [comment]
Fixed via #1309. Can you try with the latest version? I should note that it's recommended to use TLS endpoint because the SDK would buffer the whole content and calculate hash before sending out the request for non-TLS endpoint.
